imaging/ml/ml_codelab/breast_density_cloud_ml.ipynb

```diff
@@ -1,911 +1,930 @@
+    "The neural network we will use can roughly be split into two parts: \"feature extraction\" and \"classification\". In transfer learning, we take advantage of a pre-trained (checkpoint) model to do the \"feature extraction\", and add a few layers to perform the \"classification\" relevant to the specific problem. In this case, we are adding aa [dense](https://www.tensorflow.org/api_docs/python/tf/layers/dense) layer with two neurons to do the classification and a [softmax](https://www.tensorflow.org/api_docs/python/tf/nn/softmax) layer to normalize the classification score.  The mammography images will be classified as either \"2\" (scattered density) or \"3\" (heterogeneously dense). See below for diagram of the training process:\n",
+    "\n",
+    "![Inception V3](images/cloud_ml_training_pipeline.png)\n",
+    "\n",
+    "\n",
+    "The \"feature extraction\" and the \"classification\" part will be done in the following steps, respectively."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "NsamYTK9q07p"
+   },
+   "source": [
+    "### Preprocess Raw Images using Cloud Dataflow\n",
+    "\n",
+    "In this step, we will resize images to 300x300 (required for Inception V3) and will run each image through the checkpoint Inception V3 model to calculate the *bottleneck values*. This is the feature vector for the output of the feature extraction part of the model (the part that is already pre-trained). Since this process is resource intensive, we will utilize [Cloud Dataflow](https://cloud.google.com/dataflow/) in order to do this scalably. We extract the features and calculate the bottleneck values here for performance reasons - so that we don't have to recalculate them during training.\n",
+    "\n",
+    "The output of this process will be a collection of [TFRecords](https://www.tensorflow.org/guide/datasets) storing the bottleneck value for each image in the input dataset. This TFRecord format is commonly used to store Tensors in binary format for storage.\n",
+    "\n",
+    "Finally, in this step, we will also split the input dataset into *training*, *validation* or *testing*. The percentage of each can be modified using the parameters below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "wypDk298yO3l"
+   },
+   "outputs": [],
+   "source": [
+    "# GCS Bucket to store output TFRecords.\n",
+    "bottleneck_bucket = cloud_bucket_name + \"/bottleneck/\" # @param\n",
+    "\n",
+    "# Percentage of dataset to allocate for validation and testing.\n",
+    "validation_percentage = 10 # @param\n",
+    "testing_percentage = 10 # @param\n",
+    "\n",
+    "# Number of Dataflow workers. This can be increased to improve throughput.\n",
+    "dataflow_num_workers = 5 # @param\n",
+    "\n",
+    "# Staging bucket for training.\n",
+    "staging_bucket = cloud_bucket_name # @param"
```

Comment:
Do we still need to create the buckets?: 
```
!gsutil -q mb -c regional -l $location $bottleneck_bucket
!gsutil -q mb -c regional -l $location $staging_bucket
```

Reply:
no, now it's subfolders in cloud_bucket_name, so it's already created